cmd/gcp-controller-manager/readonly_resource_csr_approver.go

```diff
@@ -0,0 +1,202 @@
+			recordValidatorMetric(csrmetrics.ApprovalStatusDeny)
+			return approver.updateCSR(csr, false, challengeResult.message)
+		case challengeResult.result == true:
+			klog.Infof("validator %q: approved CSR %q", validator.name, csr.Name)
+			recordValidatorMetric(csrmetrics.ApprovalStatusApprove)
+		}
+		klog.Infof("CSR %q validation passed", csr.Name)
+	}
+
+	klog.Infof("all validators approved CSR %q", csr.Name)
+	recordMetric(csrmetrics.ApprovalStatusApprove)
+	return approver.updateCSR(csr, true, "approved by read only resource approver")
+}
+
+func (approver *readonlyResourceApprover) updateCSR(csr *capi.CertificateSigningRequest, approved bool, msg string) error {
+	if approved {
+		csr.Status.Conditions = append(csr.Status.Conditions, capi.CertificateSigningRequestCondition{
+			Type:    capi.CertificateApproved,
+			Reason:  "AutoApproved",
+			Message: msg,
+			Status:  v1.ConditionTrue,
+		})
+	} else {
+		csr.Status.Conditions = append(csr.Status.Conditions, capi.CertificateSigningRequestCondition{
+			Type:    capi.CertificateDenied,
+			Reason:  "AutoDenied",
+			Message: msg,
+			Status:  v1.ConditionTrue,
+		})
+	}
+	updateRecordMetric := csrmetrics.OutboundRPCStartRecorder("k8s.CertificateSigningRequests.updateApproval")
+	_, err := approver.ctx.client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
+	if err != nil {
+		updateRecordMetric(csrmetrics.OutboundRPCStatusError)
+		return fmt.Errorf("error updating approval status for csr: %v", err)
+	}
+	updateRecordMetric(csrmetrics.OutboundRPCStatusOK)
+	return nil
+}
+
+// -----------------------Validator-----------------------------
+type readonlyResourceCsrValidator struct {
+	name          string
+	authFlowLabel string
+
+	// challenge will challenge the request and return approved or denied
+	// (false, err) will base on err type retry or not. if it is a non-retryable error then deny the request
+	// (true, nil) will continue to next challenger. if it is the last one, will approve the request
+	// (false, nil) reject the request.
+	challenge  challengeFunc
+	permission authorization.ResourceAttributes
+}
+
+func newReadonlyResourceCsrValidator() []readonlyResourceCsrValidator {
+	return []readonlyResourceCsrValidator{
+		{
+			name:          "service account validator",
+			authFlowLabel: "readonly_resource_service_account_match",
+			challenge:     challengeServiceAccount,
+		},
+		{
+			name:          "pod annotation validator",
+			authFlowLabel: "readonly_resource_pod_annotation_valid",
+			challenge:     challengePodAnnotation,
+		},
+		{
+			name:          "rbac validator",
+			authFlowLabel: "readonly_resource_rbac_exist",
+			challenge:     challengeRbac,
+		},
+	}
+}
+
+// --------------------challenge function------------------------
+func challengeServiceAccount(content readonlyResourceCSRRequestContent) readonlyResourceCSRRequestResult {
```

Comment:
please use validate instead of challenge, the former matches the rest of the codebase better.

Reply:
done